pkg/controller/csi_handler.go

```diff
@@ -93,6 +101,99 @@ func (h *csiHandler) Init(vaQueue workqueue.RateLimitingInterface, pvQueue workq
 	h.pvQueue = pvQueue
 }
 
+// ReconcileVA lists volumes from the CSI Driver and reconciles the attachment
+// status with the corresponding VolumeAttachment object. If the attachment
+// status of the volume is different from the state on the VolumeAttachment the
+// VolumeAttachment object is patched to the correct state.
+func (h *csiHandler) ReconcileVA() error {
+	ctx, cancel := context.WithTimeout(context.Background(), h.timeout)
+	defer cancel()
+
+	// Loop over all volume attachment objects
+	vas, err := h.vaLister.List(labels.Everything())
+	if err != nil {
+		return errors.New("failed to list all VolumeAttachment objects")
+	}
+
+	// List volumes after getting volume attachment objects to prevent race
+	// condition where actual state is changed between ListVolumes and
+	// VolumeAttachment List. When listing VolumeAttachments first we can
+	// guarantee any race conditions will result in API object UPDATE failures
+	// instead of stomping on the VolumeAttachment with outdated data
```

Comment:
Is this just because both lists are RPC calls? Calling these in the other order means your actual state might be stale and get written to VA objects, while this order means your VA objects could be stale causing the UPDATE failures.

Are you sure that actual state can't become stale after you call `ListVolumes`? Or is it just a negligible problem at this point? It seems like there are zero RPC calls in the loop below, but the item could sit on the queue for some time.

Reply:
Also, do we requeue on UPDATE failures?

Reply:
Looks like the workqueue util de-dupes items.
https://github.com/kubernetes/client-go/blob/a32a6f7a3052374fff2c57c5b4b52e64dcd55eed/util/workqueue/doc.go#L21

Reply:
Also looks like this was discussed extensively above. Sorry for the repeat.

Reply:
this comment is actually out of date so I've removed it. Yeah the concurrency was discussed at length. Plopping the request to forceResync in a shared queue resolves all known race conditions